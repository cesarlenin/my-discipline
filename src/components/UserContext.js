import React, { Component } from 'react'

const UserContext = React.createContext({
    habits: [],
    actions: [],
    error: null,
    setHabits: () => {},
    setActions: () => {},
    setError: () => {},
    clearError: () => {},
  });


export default UserContext

export class UserContextProvider extends Component {
    state = {
        habits: [],
        actions: [],
        error: null,
    };
  
    setHabits = habits => {
      this.setState(habits)
    }
  
    setActions = actions => {
      this.setState(actions)
    }

    setError = error => {
        console.error(error)
        this.setState({ error })
      }
    
    clearError = () => {
        this.setState({ error: null })
    }
  
    render() {
      const value = {
        habits: this.state.habits,
        actions: this.state.actions,
        error: this.state.error,
        setError: this.setError,
        clearError: this.clearError,
        setHabits: this.setHabits,
        setActions: this.setActions,
      }
      return (
        <UserContext.Provider value={value}>
          {this.props.children}
        </UserContext.Provider>
      )
    }
  }
  