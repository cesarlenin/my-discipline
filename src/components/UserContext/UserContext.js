import React, { Component } from 'react'

const UserContext = React.createContext({
    habits: [],
    actions: [],
    authToken: null,
    error: null,
    setHabits: () => {},
    setActions: () => {},
    setAuthToken: () => {},
    setError: () => {},
    clearError: () => {},
    clearAuthToken: () => {},
    addHabit: () => {},
    addAction: () => {},
    removeHabit: () => {},
  });


export default UserContext

export class UserContextProvider extends Component {
    state = {
        habits: [],
        actions: [],
        authToken: null,
        error: null,
    };
  
    setHabits = habits => {
      this.setState({
        habits: habits
      })
    }
  
    setActions = actions => {
      this.setState({
        actions: actions
      })
    }

    setAuthToken = authToken => {
      this.setState({
        authToken: authToken
      })
    }

    setError = error => {
        console.error(error)
        this.setState({
          error: error
        })
      }
    
    clearError = () => {
        this.setState({ error: null })
    }
    clearAuthToken = () => {
      this.setState({ authToken: null })
  }

  addHabit = habits => {
    console.log(habits);
    this.setHabits([
      ...this.state.habits,
      habits
    ])
  }

  addAction = actions => {
    this.setActions([
      ...this.state.actions,
      actions
    ])
  }

  removeHabit = habitId => {
    this.setHabits(
      this.state.habits.filter((val) => val.id !== habitId)
    );
  }

  editHabit = habit => {
    this.setHabits(
      this.state.habits.map((val)=>{
        if(val.id===habit.id){
          val=habit
        }
        return val
      })
      );
  }

    render() {
      const value = {
        habits: this.state.habits,
        actions: this.state.actions,
        error: this.state.error,
        authToken: this.state.authToken,
        setError: this.setError,
        clearError: this.clearError,
        setHabits: this.setHabits,
        setActions: this.setActions,
        setAuthToken:this.setAuthToken,
        clearAuthToken:this.clearAuthToken,
        removeHabit:this.removeHabit,
        editHabit:this.editHabit,
        addHabit:this.addHabit,
      }
      return (
        <UserContext.Provider value={value}>
          {this.props.children}
        </UserContext.Provider>
      )
    }
  }
  