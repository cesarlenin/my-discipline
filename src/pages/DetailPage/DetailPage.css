.detail{
    border-bottom: 1px solid black;
    padding: 10px;
  }
  
  .hDetails{
    border-bottom: 1px solid black;
  }
  
  .result{
    font-size: x-large;
  }
