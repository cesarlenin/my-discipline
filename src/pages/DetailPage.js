import React, { Component} from 'react'
import ThreeButtons from '../components/ThreeButtons';
import UserContext from '../components/UserContext'
import WeekCalendar from '../components/WeekCalendar';

export class DetailPage extends Component {
  static contextType = UserContext;


  render() {
    const {habits, actions} = this.context;
    const shownHabit = habits.find(
      (habit) => habit.id === Number( this.props.match.params.habitId)
    );

    const datesAreOnSameDay = (first, second) =>
    first.getFullYear() === second.getFullYear() &&
    first.getMonth() === second.getMonth() &&
    first.getDate() === second.getDate();

    let progress= 0;

    let actionsShown = actions.filter(
      (action) => action.habit_id ===  Number( this.props.match.params.habitId)
      );

    for (let index = 6; index >= 0; index--) {
        let newDay= new Date();
        newDay.setDate(newDay.getDate()-index);
        let shownDate = actionsShown.find(
          (action) =>{ 
            return datesAreOnSameDay(newDay, new Date(action.date_created))
          }
        );
        if(shownDate){
          progress++
        }
    }
    
    return (
        <main>
          <header>
            <h2>{shownHabit.habit_name}</h2>
          </header>
          <section>
          <WeekCalendar id={shownHabit.id}/>
          </section>
        <section>
              <h3>description</h3>
              <p>{shownHabit.description}</p>
        </section>
        <section>
          <h3>goal</h3>
          <p>{shownHabit.goal} times a week</p>
          <br/>
          {(progress>=shownHabit.goal)? 
          <h4>congratulations you're doing great</h4>:
          <h4>keep working on it you can do it</h4>
          }
        </section>
        <section className= "aButton">
        <ThreeButtons id={shownHabit.id} {...this.props}/>
        </section>
      </main>
    )
  }
}

export default DetailPage