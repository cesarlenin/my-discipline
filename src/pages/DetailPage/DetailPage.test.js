import React from 'react';
import ReactDOM from 'react-dom';
import DetailPage from './DetailPage';
import store from '../../dummy-store';
import { BrowserRouter} from 'react-router-dom';
import UserContext from '../../components/UserContext/UserContext'
import renderer from 'react-test-renderer';

describe('<DetailPages />', () => { 

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <UserContext.Provider value={store}>
    <BrowserRouter>
  <DetailPage
  match={{params: {habitId: 1}, isExact: true, path: "", url: ""}}
  />
  </BrowserRouter>
  </UserContext.Provider>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders this UI as expected', () => {
  const tree = renderer.create(
<UserContext.Provider value={store}>
    <BrowserRouter>
  <DetailPage
  match={{params: {habitId: 1}, isExact: true, path: "", url: ""}}
  />
  </BrowserRouter>
  </UserContext.Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});

});