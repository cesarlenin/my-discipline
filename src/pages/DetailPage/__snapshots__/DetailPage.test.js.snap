// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailPages /> renders this UI as expected 1`] = `
<section>
  <header
    className="hDetails"
  >
    <h1>
       Habit Name: 
    </h1>
    <h2>
       
      Read a book
       
    </h2>
  </header>
  <div
    className="back"
  >
    <button
      onClick={[Function]}
    >
      Go Back
    </button>
  </div>
  <section
    className="detail"
  >
    <div
      className="weekCalendar"
    >
       
      <div
        className="day"
      >
        Mon
        <br />
        25
      </div>
      <div
        className="day"
      >
        Tue
        <br />
        26
      </div>
      <div
        className="day"
      >
        Wed
        <br />
        27
      </div>
      <div
        className="day"
      >
        Thu
        <br />
        28
      </div>
      <div
        className="day"
      >
        Fri
        <br />
        29
      </div>
      <div
        className="day"
      >
        Sat
        <br />
        30
      </div>
      <div
        className="day"
      >
        Sun
        <br />
        31
      </div>
    </div>
  </section>
  <section
    className="detail"
  >
    <h3>
       Description: 
    </h3>
    <p>
       
      Important
       
    </p>
  </section>
  <section
    className="detail"
  >
    <h3>
       Goal: 
    </h3>
    <p>
       
      7
       times a week 
    </p>
    <br />
    <span
      className="result red"
    >
       Keep working on it,
      <br />
       you can do it! 
    </span>
  </section>
  <section
    className="threeButtons detail"
  >
    <section
      className="buttons"
    >
      <a
        className="backGreen"
        href="/EditHabit/1"
        onClick={[Function]}
      >
        edit
      </a>
      <a
        className="backBlue"
        href="/AddAction/1"
        onClick={[Function]}
      >
        Add Action
      </a>
      <button
        className="button backRed"
        onClick={[Function]}
      >
        Delete
      </button>
    </section>
  </section>
</section>
`;
